sdk/eventhub/azure-eventhubs/azure/eventhub/receiver.py

```diff
@@ -327,3 +328,49 @@ def receive(self, max_batch_size=None, timeout=None):
             error = EventHubError("Receive failed: {}".format(e))
             self.close(exception=error)
             raise error
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close(exc_val)
+
+    def __iter__(self):
+        if not self.running:
+            self.open()
+        if not self.iter_started:
+            self.iter_started = True
+            self.messages_iter = self._handler.receive_messages_iter()
+        return self
+
+    def __next__(self):
+        while True:
+            try:
+                message = next(self.messages_iter)
+                event_data = EventData(message=message)
+                self.offset = event_data.offset
+                return event_data
+            except (errors.TokenExpired, errors.AuthenticationException):
```

Comment:
I thought we were going to refactor the exception handling into a separate function?

Reply:
Yes I would like to focus on other important stuffs for now. hierarchy, which is related to this one.
Issue 5355 is created to track this task.